derive/src/lib.rs

```diff
@@ -0,0 +1,48 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use syn::{Data, DeriveInput, Fields};
+
+#[proc_macro_derive(FromArgs)]
+pub fn derive_from_args(input: TokenStream) -> TokenStream {
+    let ast: DeriveInput = syn::parse(input).unwrap();
+
+    let gen = impl_from_args(&ast);
+    gen.to_string().parse().unwrap()
+}
+
+fn impl_from_args(input: &DeriveInput) -> TokenStream2 {
+    // FIXME: This references types using `crate` instead of `rustpython_vm`
+    //        so that it can be used in the latter. How can we support both?
+    let fields = match input.data {
+        Data::Struct(ref data) => {
+            match data.fields {
+                Fields::Named(ref fields) => fields.named.iter().map(|field| {
+                    let name = &field.ident;
+                    quote! {
+                        #name: crate::pyobject::TryFromObject::try_from_object(
+                            vm,
+                            args.take_keyword(stringify!(#name)).unwrap_or_else(|| vm.ctx.none())
```

Comment:
It probably is more efficient to iterate once through instead of each individually, and if we're writing a macro we should think about things like that.

Reply:
Once `kwargs` in `PyFuncArgs` is a map, it'll actually be doing _more_ work by iterating (O(n) vs O(1) times a constant factor). Also another issue with that approach is that you have to store a bunch of locals as `Option`s which further decreases efficiency and also just makes things more complex.